File: C#/Medium/3Sum.cs
public class Solution {
    public IList<IList<int>> ThreeSum(int[] nums) {
        // Sort the array
        Array.Sort(nums);

        var res = new List<IList<int>>();
        for (int i = 0; i < nums.Length; i++)
        {
            if (i > 0 && nums[i] == nums[i - 1])
                continue;

            int j = i + 1;
            int k = nums.Length - 1;

            while (j < k)
            {
                int sum = nums[i] + nums[j] + nums[k];

                if (sum == 0)
                {
                    res.Add(new List<int>{ nums[i], nums[j], nums[k] });
                    j++;
                    while (j < k && nums[j] == nums[j - 1])
                        j++;
                }
                else if (sum > 0)
                {
                    k--;
                }
                else
                {
                    j++;
                }
            }
        }

        return res;
    }
}

/*

    (1) Sort nums
    (2) Loop the array from start to finish with the first pointer
    (3) Implement Two Sum II algorithm on the sorted array to find triplets that sum to 0
    
    Time Complexity: O(n^2)
    Space Complexity: O(1) or O(n) (Depends on sorting implementation)
    
    Where n is the length of nums

*/

File: Swift/Medium/3Sum.swift
class Solution {
    func threeSum(_ nums: [Int]) -> [[Int]] {
        var result: [[Int]] = []
        let nums = nums.sorted()

        for i in 0..<nums.count {
            var left = i + 1
            var right = nums.count - 1

            if i > 0 && nums[i] == nums[i - 1] { continue }

            while left < right {
                let sum = nums[i] + nums[left] + nums[right]
                if sum > 0 {
                    right -= 1
                }
                else if sum < 0 {
                    left += 1
                }
                else {
                    result.append([nums[i], nums[left], nums[right]])
                    left += 1
                    while (nums[left] == nums[left - 1] && left < right) {
                        left += 1
                    }
                }
            }
        }

        return result
    }
}

/*

    Solution:
    * Sort the input array
    * Run a two-pointer find for two numbers that make the sum 0

    Time: O(n^2)
    Space: O(n) or O(logn) depending on the sorting algorithm

    Note: Ignoring the output array for the purpose of complexity analysis

*/
